Update the spec to fix the issue 782 (#943)
jakartaee/data: spec/src/main/asciidoc/query-language.asciidoc

The characters `(`, `)`, and `,` are _punctuation characters_.

[WARNING]
====

> When working with NoSQL databases, the support for arithmetic operations and the use of punctuation characters like parentheses for precedence may vary significantly:

Reviewer: ```suggestion
When working with NoSQL databases, the support for arithmetic operations and support of parentheses for precedence might vary significantly:
```

- **Key-Value Databases**: Arithmetic operations (`+`, `-`, `*`, `/`) are not supported. These databases are designed for simple key-based lookups and lack query capabilities for complex operations.

- **Wide-Column Databases**: Arithmetic operations are not required to be supported. Some wide-column databases might offer limited support, which might require secondary indexing even for basic querying.

> - **Document and Graph Databases**: These databases typically support arithmetic operations and punctuation characters for precedence. However, the exact behavior and extent of support can vary significantly between providers.

Reviewer: ```suggestion
- **Document Databases**: Support of arithmetic operations and support of parenthesis for precedence are not required, although databases typically offer these capabilities. Behavior and extent of support can vary significantly between providers.

- **Graph Databases**: Support of parentheses for precedence is required. Support for arithmetic operations is not required but is typically offered by databases. Behavior and extent of support can vary significantly between providers.
```

> - **Document and Graph Databases**: These databases typically support arithmetic operations and punctuation characters for precedence. However, the exact behavior and extent of support can vary significantly between providers.

Author: We need to update TCK.

Graph Context:

We do have two standard languages on Graph:

- [OpenCypher](https://opencypher.org/): This comes from the most popular Graph NoSQL database, Neo4J. This one supports parentheses.
- [Apache Tinkerpop](https://tinkerpop.apache.org/): This one belongs to the Apache foundation and does not support those operations as OpenCypher; instead, it focuses more on the navigation of the graph exploring navigation and queries on the relationship.
- Unlike SQL, even those standards are not mandatory for Graph databases; those are just popular. For example, ArangoDB uses a different approach for Graph Query: [its query language](https://docs.arangodb.com/3.13/aql/graphs/traversals/).

⚠️ I hope not to be boring with that context

> - **Document and Graph Databases**: These databases typically support arithmetic operations and punctuation characters for precedence. However, the exact behavior and extent of support can vary significantly between providers.

Reviewer: > Thus, in theory, it should, but in practice, I don't have this guarantee.

If we are unaware of any Graph databases that will have a problem with it, I think it would be fine to keep the requirement in place.  We would find out from the TCK if it ever become a problem and could opt to relax it at that point if necessary.

> Due to the diversity of NoSQL database types and their querying capabilities, there is no guarantee that all NoSQL providers will support punctuation characters such as parentheses `(`, `)` for defining operation precedence. It is recommended to consult your NoSQL provider's documentation to confirm the supported query features and their behavior.

Author: The same as here: https://github.com/jakartaee/data/pull/943/files#r1924688232
We need to update the TCK.

NOTE: Portability is maximized when Jakarta Data providers interpret equality and inequality operators in a manner consistent with the implementation of `Object.equals()` or `Comparable.compareTo()` for the assigned Java type.

[WARNING]
====
When using NoSQL databases, there are limitations to the support of equality and inequality operators:

> 1. **Key-Value Databases**: Support for equality and inequality restrictions on the key attribute is required. The key attribute is defined by the annotation `jakarta.nosql.Id`. Key-value databases are not required to support any other restrictions.

Reviewer: Per the information that `Not` isn't required of Key-Value,
```suggestion
1. **Key-Value Databases**: Support for the equality restriction on the key attribute is required. The key attribute is defined by the annotation `jakarta.nosql.Id`. Key-value databases are not required to support any other restrictions.
```

- This specification does not define an order for UUID values, which is provider-dependent.

[WARNING]
====
When using NoSQL databases, sorting support varies by database type:

- **Key-Value Databases**: Sorting of results is not supported.

- **Wide-Column Databases**: Support for sorting of results is not required. In general, sorting is not natively supported. When sorting is available, it is typically limited to:
  * The key attribute, defined by an annotation such as `jakarta.nosql.Id`.
  * Fields that are indexed as secondary indexes.

> - **Graph and Document Databases**: Sorting is generally supported, especially for single fields. However, behavior for compound sorting (sorting by multiple fields) may vary due to:
> - Potential instability with tied values, where sorting results for equivalent values may differ across queries.
> - Schema flexibility and mixed data types, which may affect consistent sorting behavior.
> - Dependence on indexes and internal storage order, requiring proper indexing to ensure predictable sorting.
> - The distributed nature of sharded clusters, where sorting across shards may introduce additional complexity.

Reviewer: ```suggestion
- **Graph and Document Databases**: Support for sorting by a single entity attribute is required. Support for compound sorting (sorting by multiple entity attributes) is not required and may vary due to:
  * Potential instability with tied values, where sorting for equivalent values may differ across queries.
  * Schema flexibility and mixed data types.
  * Dependence on indexes and internal storage order, requiring proper indexing to ensure predictable sorting.
  * The distributed nature of sharded clusters, where sorting across shards may introduce additional complexity.
```

Syntactically, logical operators are parsed with lower precedence than <<Equality and inequality operators,equality and inequality operators>> and other <<Conditional expressions,conditional expressions listed above>>. The `not` operator has higher precedence than `and` and `or`. The `and` operator has higher precedence than `or`.

[WARNING]
====
When using NoSQL databases, the support for restrictions varies depending on the database type:

> - **Key-Value Databases**: These databases will not support the described restrictions. Queries in key-value databases are typically limited to operations using the key attribute only.

Reviewer: ```suggestion
- **Key-Value Databases**: Support for the equality restriction is required for the `Id` attribute. There is no requirement to support other types of restrictions or restrictions on other entity attributes.
```

> - **Wide-Column Databases**: Wide-column databases typically support only the `AND` operator for conditions, primarily focusing on the key attribute annotated with `jakarta.nosql.Id`. Non-key attributes can be used if they are indexed as secondary indexes, which is important for compatibility, although support varies by database provider.

Reviewer: ```suggestion
- **Wide-Column Databases**: Wide-column databases are not required to support the `AND` operator or the `OR` operator. Restrictions must be supported for the key attribute that is annotated with `jakarta.nosql.Id`. Support for restrictions on other attributes is not required. Typically they can be used if they are indexed as secondary indexes, although support varies by database provider.
```

> - **Graph and Document Databases**: These databases are expected to support all the restrictions described in this section. However, precedence between `And` and `Or` operators is not guaranteed unlike relational databases. The behavior of precedence may vary significantly based on the NoSQL provider.

Reviewer: ```suggestion
- **Graph and Document Databases**: The `AND` and `OR` operators and all of the restrictions described in this section must be supported. Precedence between `AND` and `OR` operators is not guaranteed and may vary significantly based on the NoSQL provider.
```

The `select` clause is optional in `select` statements. When it is missing, the query returns the queried entity.

[WARNING]
====
When working with NoSQL databases, the `select` clause behavior may vary depending on the database structure and capabilities:

> - **Key-Value Databases**: These databases generally do not support `select` clauses beyond retrieving values by their keys. Complex path expressions or aggregate functions like `count(this)` are typically unsupported.

Reviewer: ```suggestion
- **Key-Value Databases**: These databases generally do not support `select` clauses beyond retrieving values by their keys. Support for complex path expressions and aggregate functions like `count(this)` is not required.
```

> - **Wide-Column Databases**: The ability to use a `select` clause may depend on the presence of secondary indexes. Without secondary indexes, selection is often restricted to key-based operations.

Reviewer: ```suggestion
- **Wide-Column Databases**: The ability to use a `select` clause may depend on the presence of secondary indexes. Without secondary indexes, selection is often restricted to key-based operations. Support for `count(this)` is not required.
```

> - **Graph and Document Databases**: These databases usually support flexible `select` clauses, including path expressions and aggregate functions like `count(this)`. However, performance may vary based on the size and indexing of the dataset.

Reviewer: ```suggestion
- **Graph and Document Databases**: Support for flexible `select` clauses, including path expressions and aggregate functions like `count(this)` is required. Performance might vary based on the size and indexing of the dataset.
```
